enforcer-rules/src/main/java/org/apache/maven/plugins/enforcer/DependencyConvergence.java

```diff
@@ -51,58 +46,41 @@ public class DependencyConvergence implements EnforcerRule {
-            MavenSession session = (MavenSession) helper.evaluate("${session}");
-            DependencyCollectorBuilder dependencyCollectorBuilder =
-                    helper.getComponent(DependencyCollectorBuilder.class);
-            ArtifactRepository repository = (ArtifactRepository) helper.evaluate("${localRepository}");
-
-            ProjectBuildingRequest buildingRequest =
-                    new DefaultProjectBuildingRequest(session.getProjectBuildingRequest());
-            buildingRequest.setProject(project);
-            buildingRequest.setLocalRepository(repository);
-            ArtifactFilter filter = (Artifact a) ->
-                    ("compile".equalsIgnoreCase(a.getScope()) || "runtime".equalsIgnoreCase(a.getScope()))
-                            && !a.isOptional();
-
-            return dependencyCollectorBuilder.collectDependencyGraph(buildingRequest, filter);
-        } catch (ExpressionEvaluationException | ComponentLookupException e) {
-            throw new EnforcerRuleException("Unable to lookup a component " + e.getLocalizedMessage(), e);
-        } catch (DependencyCollectorBuilderException e) {
-            throw new EnforcerRuleException("Could not build dependency tree " + e.getLocalizedMessage(), e);
-        }
-    }
-
     @Override
     public void execute(EnforcerRuleHelper helper) throws EnforcerRuleException {
         if (log == null) {
             log = helper.getLog();
         }
         try {
-            DependencyNode node = getNode(helper);
-            DependencyVersionMap visitor = new DependencyVersionMap(log);
-            visitor.setUniqueVersions(uniqueVersions);
-            node.accept(visitor);
-            List<CharSequence> errorMsgs = new ArrayList<>();
-            errorMsgs.addAll(getConvergenceErrorMsgs(visitor.getConflictedVersionNumbers(includes, excludes)));
+            DependencyNode node = ArtifactUtils.resolveTransitiveDependencies(
+                    helper,
+                    // TODO: use a modified version of ExclusionDependencySelector to process excludes and includes
+                    new DependencySelector() {
+                        @Override
+                        public boolean selectDependency(Dependency dependency) {
+                            // regular OptionalDependencySelector only discriminates optional dependencies at level 2+
+                            return !dependency.isOptional()
+                                    // regular ScopeDependencySelector is case-sensitive
+                                    && !dependency.getScope().equalsIgnoreCase(Artifact.SCOPE_TEST);
+                        }
+
+                        @Override
+                        public DependencySelector deriveChildSelector(DependencyCollectionContext context) {
+                            return this;
+                        }
+                    },
+                    new ExclusionDependencySelector());
```

Comment:
Why we need here special selectors?

provided scope is missing for exclusion - see MENFORCER-407 

https://github.com/apache/maven/blob/master/maven-resolver-provider/src/main/java/org/apache/maven/repository/internal/MavenRepositorySystemUtils.java

Reply:
> Why we need here special selectors?

It's explained in the comment: // regular OptionalDependencySelector only discriminates optional dependencies at level 2+

Reply:
the scope selector was not passing tests if the scope case was wrong; but I'll apply toLowerCase in ArtifactUtils.toArtifact(DependencyNode)

Reply:
Changed to:

                    new OptionalDependencySelector(),
                    new ScopeDependencySelector(SCOPE_TEST, SCOPE_PROVIDED),
                    new ExclusionDependencySelector());

Reply:
Nope:

[INFO]   The build exited with code 1. See /Users/ajarmoniuk/IdeaProjects/maven-enforcer/maven-enforcer-plugin/target/it/dependencies_not_converging_test_scope/build.log for details.
[INFO]           dependencies_not_converging_test_scope/pom.xml ... FAILED (2.0 s)

Reply:
will check why

Reply:
@slawekjaranowski unfortunately, out of the box selectors only work on transitive dependencies: direct dependencies are always included. That was the primary reason I used this, I'll put a comment about it.